### PR TITLE
Add :http_verbs option

### LINKED ISSUE #51: Crumbs are never current on POST
### Describe the problem
Crumbs are never current when rendered on a response to e.g. an HTTP Post.

### Steps to reproduce the problem
Render the same crumbs in response to HTTP POST and GET request.

### Actual behaviour
On GET, some crumbs can be current, and others not.
On POST, crumbs are never current.

### Expected behaviour
Respect the behaviour configured using :match.

### Describe your environment
* OS version: MacOS Big Sur
* Ruby version: 2.7
* Loaf version: 0.10

This line here might be reasonably blunt:
https://github.com/piotrmurach/loaf/blob/v0.10.0/lib/loaf/view_extensions.rb#L69

Reckons? Thanks!

### PR DESCRIPTION
**Pull request by hoppergee**
### Describe the change
Add `:http_verbs` option to breadcrumb parameters

### Why are we doing this?
When I use this gem on some legacy projects, I met a circumstance which I have to show breadcrumbs on POST/PATCH action. I tried to add an option `:http_only` at #49. But @piotrmurach suggest me to use `http_verbs`.

### Benefits
Now the current breadcrumb can match with multiple HTTP methods

```ruby
breadcrumb 'Step 1', onboard_step_path(step: 1), http_verbs: [:get]
breadcrumb 'Step 2', onboard_step_path(step: 2), http_verbs: [:get, :post, :delete]
breadcrumb 'Step 3', onboard_step_path(step: 3), http_verbs: :all
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -207,6 +208,29 @@ To make a breadcrumb current based on the query parameters do:
...
+The `:http_verbs` key accepts `:all` or an array with following values:
+
+* `:get`
+* `:post`
+* `:put`
+* `:patch`
+* `:delete`
+* `:head`
+
+Its default value is `[:get]`
```
Comment (piotrmurach) [inline README.md]: For backwards compatibility, I'd include `:head` as well.

File: README.md [hunk 2]
```diff
@@ -207,6 +208,29 @@ To make a breadcrumb current based on the query parameters do:
...
+**Loaf** allows you to match on multiple HTTP verbs in order to make a breadcrumb current with the `:http_verbs` option.
+
+The `:http_verbs` key accepts `:all` or an array with following values:
+
+* `:get`
+* `:post`
+* `:put`
+* `:patch`
+* `:delete`
+* `:head`
```
Comment (piotrmurach) [inline README.md]: Could you please include `:options`, `:link`, `:unlink` and `:trace` for completeness and compatibility with `rack`?

File: lib/loaf/configuration.rb [hunk 3]
```diff
@@ -13,6 +14,8 @@ class Configuration
 
     DEFAULT_MATCH = :inclusive
 
+    DEFAULT_HTTP_VERBS = [:get]
```
[the remark this answers is not in the document]
Comment (piotrmurach) [reply lib/loaf/configuration.rb]: Please add `:head` and convert to frozen symbol array:

File: lib/loaf/view_extensions.rb [hunk 4]
```diff
@@ -65,8 +69,8 @@ def breadcrumb_trail(options = {})
     #   the pattern to match on
     #
     # @api public
-    def current_crumb?(path, pattern = :inclusive)
-      return false unless request.get? || request.head?
+    def current_crumb?(path, pattern = :inclusive, http_verbs = [:get])
```
Comment (piotrmurach) [inline lib/loaf/view_extensions.rb]: Please convert `http_verbs` from positional argument to keyword argument and reference `Loaf.configuration.http_verbs`.

File: Gemfile [hunk 5]
```diff
@@ -8,6 +8,7 @@ gem 'yard',        '~> 0.9.24'
 gem 'capybara',    '~> 3.30.0'
 gem 'rspec-rails', '~> 3.9.0'
 gem 'public_suffix', '~> 2.0.5'
+gem 'byebug'
```
[the remark this answers is not in the document]
Comment (piotrmurach) [reply Gemfile]: Please remove `byebug` as a dependency. The Gemfile is used to generate all the gemfiles for different Rails versions and it only should contain 'essential' gems for testing etc. I don't believe we should be 'enforcing' any debugging gems in this project.

File: README.md [hunk 6]
```diff
@@ -207,6 +208,29 @@ To make a breadcrumb current based on the query parameters do:
...
+
+Its default value is `[:get]`
+
+For example:
+
+```ruby
+http_verbs: %w[get head]
+http_verbs: %w[post get head]
```
Comment (piotrmurach) [inline README.md]: Since request methods are described with symbols, it'd be more appropriate and probably desirable to use symbol arrays here:

```ruby
http_verbs: %i[get head]
http_vers: %i[post get head]
```

File: lib/loaf/configuration.rb [hunk 7]
```diff
@@ -4,7 +4,8 @@ module Loaf
   class Configuration
     VALID_ATTRIBUTES = [
       :locales_path,
-      :match
+      :match,
+      :http_verbs
```
Comment (piotrmurach) [inline lib/loaf/configuration.rb]: I'd keep options in alphabetical order.

**Review by piotrmurach — COMMENTED** [review]
This is looking very promising! Thanks for a quick turnaround on this issue. 🙏   I left few comments. Could you please also use double-quoted strings everywhere? Sorry for the confusion, there is a mixture of both styles but going forward I'd like to keep them double-quoted.

File: lib/loaf/configuration.rb [hunk 8]
```diff
@@ -13,6 +14,8 @@ class Configuration
 
     DEFAULT_MATCH = :inclusive
 
+    DEFAULT_HTTP_VERBS = [:get, :head].freeze
```
[the remark this answers is not in the document]
Comment (piotrmurach) [reply lib/loaf/configuration.rb]: Would it be possible to use a symbol array instead?

**Review by piotrmurach — COMMENTED** [review]
Things are looking good! Would you please address the remaining 'houndci' style issues? There are still some single-quoted strings left, arrays that can be converted to symbol arrays etc. Also, there are some tests failures in Rails `3.2`. I know it's old Rails - sorry! But I don't want to break things in this release. I will consider dropping support later on in a separate release. Thank you!

File: spec/integration/breadcrumb_trail_spec.rb [hunk 9]
```diff
@@ -70,4 +70,54 @@
       )
     end
   end
+
+  it "match to Non-GET methods" do
```
[the remark this answers is not in the document]
Comment (hoppergee) [reply spec/integration/breadcrumb_trail_spec.rb:74]: This what I want to test. And it is simple enough. I'm ok with this.

File: lib/loaf/view_extensions.rb [hunk 10]
```diff
@@ -128,5 +132,17 @@ def _expand_url(url)
...
+    #
+    # @retun [Boolean]
+    #
+    # @api private
+    def match_http_verbs(http_verbs)
+      http_verbs ||= Loaf.configuration.http_verbs
+      return true if http_verbs == :all
+
+      http_verbs.any? { |verb| request.try("#{verb}?") }
+    end
```
Comment (piotrmurach) [inline lib/loaf/view_extensions.rb]: I like this refactoring! 👍

File: spec/integration/breadcrumb_trail_spec.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (piotrmurach) [reply spec/integration/breadcrumb_trail_spec.rb:74]: I'm not that much concerned with block length in specs. Agreed let's ignore.

File: spec/integration/breadcrumb_trail_spec.rb [hunk 11]
```diff
@@ -70,4 +70,54 @@
...
+    click_on "Save & Next" # PUT
+    expect(page).to have_selector("h1", text: "Step 3")
+    page.within "#breadcrumbs .selected" do
+      expect(page.html).to include('<a href="/onboard/step/3">Step 3</a>')
+    end
+
+    if Rails.version >= '4.0.0'
+      click_on "Save & Next" # PATCH
+      expect(page).to have_selector("h1", text: "Step 4")
+      page.within '#breadcrumbs .selected' do
```
[the remark this answers is not in the document]
Comment (piotrmurach) [reply spec/integration/breadcrumb_trail_spec.rb]: Sorry, small nitpick but let's convert this one to double-strings.

**Review by piotrmurach — COMMENTED** [review]
Thank you Hopper for making these changes! Things are looking really good. There is a small style issue left. 

There is one dilemma that I'm 'debating' which is the name of the parameter `:http_verbs`. The `verbs` term is kind of inaccurate because these are not all verbs and also it is rather 'colloquial' usage. The [HTTP rfc](https://datatracker.ietf.org/doc/html/rfc7231#page-21) talks about HTTP methods. It feels more appropriate to use `methods`. I'm equally debating whether it should be `request_methods` or `http_methods`? What're your thoughts on this? I want to make sure we pick the best configuration name.

**Comment by hoppergee** [comment]
> Thank you Hopper for making these changes! Things are looking really good. There is a small style issue left.
> 
> There is one dilemma that I'm 'debating' which is the name of the parameter `:http_verbs`. The `verbs` term is kind of inaccurate because these are not all verbs and also it is rather 'colloquial' usage. The [HTTP rfc](https://datatracker.ietf.org/doc/html/rfc7231#page-21) talks about HTTP methods. It feels more appropriate to use `methods`. I'm equally debating whether it should be `request_methods` or `http_methods`? What're your thoughts on this? I want to make sure we pick the best configuration name.

Maybe the `request_methods` is better, as Rack is using `request_method`.

https://github.com/rack/rack/blob/d15dd728440710cfc35ed155d66a98dc2c07ae42/lib/rack/request.rb#L174-L202

File: README.md [hunk 12]
```diff
@@ -207,6 +208,33 @@ To make a breadcrumb current based on the query parameters do:
...
+* `:put`
+* `:patch`
+* `:delete`
+* `:head`
+* `:options`
+* `:link`
+* `:unlink`
+* `:trace`
+
+Its default value is `%i[get head]`
```
Comment (piotrmurach) [inline README.md]: I'd reword to 

It defaults to`%i[get head]`.

File: README.md [hunk 13]
```diff
@@ -207,6 +208,33 @@ To make a breadcrumb current based on the query parameters do:
...
+
+For example:
```
Comment (piotrmurach) [inline README.md]: How about?

Here are some examples:

File: lib/loaf/view_extensions.rb [hunk 14]
```diff
@@ -65,8 +69,8 @@ def breadcrumb_trail(options = {})
     #   the pattern to match on
     #
     # @api public
-    def current_crumb?(path, pattern = :inclusive)
-      return false unless request.get? || request.head?
+    def current_crumb?(path, pattern = :inclusive, request_methods: nil)
```
Comment (piotrmurach) [inline lib/loaf/view_extensions.rb]: How about using `Loaf.configuration.request_methods` in place of `nil`?

File: lib/loaf/view_extensions.rb [hunk 15]
```diff
@@ -128,5 +132,17 @@ def _expand_url(url)
...
       end
     end
+
+    # Check if the HTTP request methods are allowed
+    #
+    # @retun [Boolean]
+    #
+    # @api private
+    def match_request_methods(request_methods)
+      request_methods ||= Loaf.configuration.request_methods
```
Comment (piotrmurach) [inline lib/loaf/view_extensions.rb]: I'd consider moving this as the default value in the `current_crumb?`.

File: spec/integration/breadcrumb_trail_spec.rb [hunk 16]
```diff
@@ -70,4 +70,54 @@
...
+
+    expect(page).to have_selector("h1", text: "Onboard")
+    page.within "#breadcrumbs" do
+      expect(page).to have_content("Onboard")
+    end
+
+    click_link "Step 1" # GET
+    expect(page).to have_selector("h1", text: "Step 1")
+    page.within "#breadcrumbs" do
+      expect(page.html).to include('<a href="/onboard">Onboard</a>')
```
Comment (piotrmurach) [inline spec/integration/breadcrumb_trail_spec.rb]: I'd change all similar checks to use capybara helper instead:

```ruby
expect(page).to have_link("Onboard", href: "/onboard")
```

File: spec/integration/breadcrumb_trail_spec.rb [hunk 17]
```diff
@@ -70,4 +70,54 @@
...
+    end
+    page.within "#breadcrumbs .selected" do
+      expect(page.html).to include('<a href="/onboard/step/1">Step 1</a>')
```
Comment (piotrmurach) [inline spec/integration/breadcrumb_trail_spec.rb]: Similar to previous comment please consider using capybara helper:

```ruby
expect(page).to have_link("Step 1", href: "/onboard/step/1")
```

File: spec/integration/breadcrumb_trail_spec.rb [hunk 18]
```diff
@@ -70,4 +70,54 @@
       )
     end
   end
+
+  it "match to Non-GET methods" do
+    visit onboard_path
+
+    expect(page).to have_selector("h1", text: "Onboard")
+    page.within "#breadcrumbs" do
```
Comment (piotrmurach) [inline spec/integration/breadcrumb_trail_spec.rb]: I'd replace all instances of  `page.within` and use `within` directly:

```ruby
within "#breadcrumbs" do
```

File: spec/integration/breadcrumb_trail_spec.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (piotrmurach) [reply spec/integration/breadcrumb_trail_spec.rb]: Still seems to be single-quoted.

File: spec/unit/view_extensions/breadcrumb_trail_spec.rb [hunk 19]
```diff
@@ -280,4 +280,34 @@
     expect(view.breadcrumb_trail.map(&:url)).to eq(%w[/ /posts /posts/1])
     expect(view.breadcrumb_trail(match: :exact).map(&:current?)).to eq([false, false, true])
   end
+
+  it "match current path with :request_methods" do
```
Comment (piotrmurach) [inline spec/unit/view_extensions/breadcrumb_trail_spec.rb]: I'd describe specifically what happens in a test, for example: 

```ruby
it "matches the current path with :request_methods set to custom"
```

File: spec/unit/view_extensions/breadcrumb_trail_spec.rb [hunk 20]
```diff
@@ -280,4 +280,34 @@
...
+    view = DummyView.new
+    view.breadcrumb("posts", "/posts")
+    view.set_path("/posts")
+    view.set_request_method(:post)
+
+    trail = view.breadcrumb_trail.map(&:to_a)
+    expect(trail).to eq([["posts", "/posts", false]])
+  end
+
+  it "match current path with :request_methods => :all" do
```
Comment (piotrmurach) [inline spec/unit/view_extensions/breadcrumb_trail_spec.rb]: Possibly reword to:

```ruby
it "matches current path with :request_methods set to :all"
```

File: spec/unit/view_extensions/breadcrumb_trail_spec.rb [hunk 21]
```diff
@@ -280,4 +280,34 @@
...
+    view = DummyView.new
+    view.breadcrumb("posts", "/posts", request_methods: %i[get post])
+    view.set_path("/posts")
+    view.set_request_method(:post)
+
+    trail = view.breadcrumb_trail.map(&:to_a)
+    expect(trail).to eq([["posts", "/posts", true]])
+  end
+
+  it "fail to match current path with :request_methods" do
```
Comment (piotrmurach) [inline spec/unit/view_extensions/breadcrumb_trail_spec.rb]: Consider rewording to:

```ruby
it "fails to match current path with :request_methods set to default"
```

**Review by piotrmurach — COMMENTED** [review]
Apologies for not having replied to your request for review earlier. I took a break from open-source. I left a few comments. Let's move this PR forward.

**Comment by hoppergee** [comment]
> Apologies for not having replied to your request for review earlier. I took a break from open-source. I left a few comments. Let's move this PR forward.

Glad to hear that! I'll get it done sometime this week.

**Comment by hoppergee** [comment]
I just fixed all those style issues. Also fixed a wrong proc config test issue

**Comment by ankalus** [comment]
Are there plans to merge these changes?
